src/Features/LanguageServer/Protocol/Handler/SemanticTokens/SemanticTokensHelpers.cs

```diff
@@ -162,6 +162,41 @@ static SemanticTokensHelpers()
             return (ComputeTokens(text.Lines, updatedClassifiedSpans, tokenTypesToIndex), isFinalized);
         }
 
+        private static async Task GetClassifiedSpansForDocumentAsync(
+            Document document,
+            TextSpan textSpan,
+            ArrayBuilder<ClassifiedSpan> classifiedSpans,
+            bool isRazorDoc,
+            CancellationToken cancellationToken)
+        {
+            var classificationService = document.GetRequiredLanguageService<IClassificationService>();
+
+            // Case 1 - Generated Razor documents:
+            //     In Razor, the C# syntax classifier does not run on the client. This means we need to return both
+            //     syntactic and semantic classifications.
+            // Case 2 - C# and VB documents:
+            //     In C#/VB, the syntax classifier runs on the client. This means we only need to return semantic
+            //     classifications.
+            //
+            // Ideally, Razor will eventually run the classifier on their end so we can get rid of this special
+            // casing: https://github.com/dotnet/razor-tooling/issues/5850
+            if (isRazorDoc)
+            {
+                var spans = await ClassifierHelper.GetClassifiedSpansAsync(
+                    document, textSpan, cancellationToken, removeAdditiveSpans: false, fillInClassifiedSpanGaps: false).ConfigureAwait(false);
```

Comment:
might be good to make a note of why the flags are set this way for LSP.  I saw a mention below for removeAdditiveSpans, but not for fillInClassifiedSpanGaps